### ISSUE TITLE
Sign up Form can't be closde by clicking in the background 

### ISSUE THREAD

**Issue by sohaibMan** [AntonioErdeljac/next13-airbnb-clone#3]
I also can't scroll the top to click the close button
![Screenshot from 2023-04-02 16-39-46](https://user-images.githubusercontent.com/96356071/229366606-2af83588-14c1-40e5-b3a6-48c40fe9c5f9.png)

**Comment by shyakadavis** [comment]
Hi, @sohaibMan 

I wanted to share a work-around to that issue. I used a similar approach on the UserMenu, but didn't want to do so for the Auth modals. But it works either way.

First, create two (2) references using the `useRef` hook. These will help us reference our elements in the DOM.

```
 	const buttonRef = useRef<HTMLButtonElement>(null);
 	const modalRef = useRef<HTMLDivElement>(null);
  ```
  
  Header over to the button you are using to trigger the modal display, and add the first reference, i.e `buttonRef`
  
  ```
  // ⚠️ Notice that for me I am using semantic html, and I use buttons instead of plain divs for cases where I know that that element behaves as a button.
   <button
          ref={buttonRef}
          type="button"
          onClick={toggleOpen}
          className="..."
        >
       ...
   </button>
  ```
  
  After that, head over the actual modal, and target the parent `div` of what is actually being rendered, and add the second ref.
  
  ```
  <div
          ref={modalRef}
          className="..."
        >
  </div>
 ```
        
From here, what we need is way to track certain events, and act accordingly.

We can use the `useEffect` hook, to track the MouseEvent for when a user clicks outside the modal.

```
  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (
        buttonRef.current &&
        modalRef.current &&
        !buttonRef.current.contains(e.target as Node) &&
        !modalRef.current.contains(e.target as Node)
      ) {
        setIsOpen(false);
      }
    };

	 // this is not necessary, but it is a nice addition to handle keyboard events as well
    const handleCloseModal = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        setIsOpen(false);
      }
    };

	// finally, add the event listeners to the DOM.
    document.addEventListener('mousedown', handleClickOutside);
    document.addEventListener('keydown', handleCloseModal);

	// ⚠️ remember to remove the listeners from useEffect in order to avoid any leaking.
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
      document.removeEventListener('keydown', handleCloseModal);
    };
  }, []);
```

Hope that helps.

**Comment by sohaibMan** [comment]
@shyakadavis Yeah it helps , I did the same thing with another apraoch , check the PR I made for further explanation

**Comment by gurbaaz27** [comment]
@sohaibMan were you able to solve the scrolling issue in Modal? This creates a lot of problem when using the webapp on mobile, since Modal takes up entire screen, and then it is impossible to scroll up to close the Modal.

@AntonioErdeljac any help here would be appreciated!

**Comment by sohaibMan** [comment]
> @sohaibMan were you able to solve the scrolling issue in Modal? This creates a lot of problem when using the webapp on mobile, since Modal takes up entire screen, and then it is impossible to scroll up to close the Modal.
> 
> @AntonioErdeljac any help here would be appreciated!

yeah,check the pr I've made it's a simple solution

**Comment by sohaibMan** [comment]
> > @sohaibMan were you able to solve the scrolling issue in Modal? This creates a lot of problem when using the webapp on mobile, since Modal takes up entire screen, and then it is impossible to scroll up to close the Modal.
> > @AntonioErdeljac any help here would be appreciated!
> 
> Yeah, check the pr number #4   I've made,it is a simple solution

**Comment by gurbaaz27** [comment]
@sohaibMan yes I checked that..
I believe that only deals with clicking on background to close the modal,, and not help in the scrolling issue in modal.

**Comment by sohaibMan** [comment]
> @sohaibMan yes I checked that.. I believe that only deals with clicking on the background to close the modal,, and not help in the scrolling issue in the modal.

Yeah exactly, you can add some media query to the model  to adjust the width  or  add a close button

**Comment by hongk-bb** [comment]
```
/*Modal.tsx*/

{/*content*/}
  <div
    className={`
    md:mt-36
    duration-300
    h-full
    ${showModal ? "translate-y-0" : "translate-y-full"}
    ${showModal ? "opacity-100" : "opacity-0"}
  `}
  >
```
I tried to use the media query to add "md:mt-36" to the modal box. I'm not sure if this is the best way, but it helps me.

**Comment by hongk-bb** [comment]
```
/* When the height of Modal exceeds the height of the screen, set margin-top */
const ModalRef = useRef<HTMLDivElement>(null)
const handleModalMarginTop = useCallback(() => {
  if (ModalRef.current) {
    const viewportHeight = window.innerHeight
    const modalHeight = ModalRef.current.clientHeight

    if (modalHeight > viewportHeight) {
      ModalRef.current.style.marginTop = `
        calc(${modalHeight - viewportHeight}px + 3rem)
      `
    }
  }
}, [ModalRef.current])
useEffect(() => {
  handleModalMarginTop()
}, [handleModalMarginTop])

if (!isOpen) {
  return null
}

return (...
```

```
{/*content*/}
    <div
      ref={ModalRef}
      className={`
        translate
        duration-300
        h-full
        ${showModal ? "translate-y-0" : "translate-y-full"}
        ${showModal ? "opacity-100" : "opacity-0"}
      `}
```

Illustration : 
![35499FCEA6D3DC77E34C3EADC966EFFB](https://github.com/AntonioErdeljac/next13-airbnb-clone/assets/110291262/f93d7647-a4dc-4457-b39a-1243f885a9be)

1. The browser's viewport is 100vh, but our Modal exceeds it due to too many child elements.
2. However, it is still centered, so my idea is to add a margin top and move it so that the top and bottom margins are symmetrical.
3. I cannot guarantee that the picture I drew is completely correct, but it does work for different Modals

Achievement:

![image](https://github.com/AntonioErdeljac/next13-airbnb-clone/assets/110291262/87c75523-9758-468d-9b89-32cf9cd8f5b6)


This is just a way of thinking, hoping to be helpful to everyone.

**Comment by gurbaaz27** [comment]
@hongk-bb thx a lot man!
